pinot-core/src/main/java/com/linkedin/pinot/core/minion/SegmentConverter.java

```diff
@@ -0,0 +1,232 @@
+
+    _recordAggregator = recordAggregator;
+    _aggregationColumns = aggregationColumns;
+    _indexingConfig = indexingConfig;
+  }
+
+  public List<File> convertSegment() throws Exception {
+    List<File> resultFiles = new ArrayList<>();
+    for (int currentPartition = 0; currentPartition < _totalNumPartition; currentPartition++) {
+      // Mapping stage
+      Preconditions.checkNotNull(_recordTransformer);
+      String mapperOutputPath = _workingDir.getPath() + File.separator + MAPPER_PREFIX + currentPartition;
+      try (MapperRecordReader mapperRecordReader = new MapperRecordReader(_inputIndexDirs, _recordTransformer,
+          _recordPartitioner, _totalNumPartition, currentPartition)) {
+        buildSegment(mapperOutputPath, _tableName, _segmentName, mapperRecordReader, null);
+      }
+      File outputSegment = new File(mapperOutputPath + File.separator + _segmentName);
+
+      // Sorting on aggregation column & reduce
+      if (_recordAggregator != null && _aggregationColumns != null && _aggregationColumns.size() > 0) {
+        String reducerOutputPath = _workingDir.getPath() + File.separator + REDUCER_PREFIX + currentPartition;
+        try (ReducerRecordReader reducerRecordReader = new ReducerRecordReader(outputSegment, _recordAggregator,
+            _aggregationColumns)) {
+          buildSegment(reducerOutputPath, _tableName, _segmentName, reducerRecordReader, null);
+        }
+        outputSegment = new File(reducerOutputPath + File.separator + _segmentName);
+      }
+
+      // Sorting on sorted column and creating indices
+      if (_indexingConfig != null) {
+        List<String> sortedColumn = _indexingConfig.getSortedColumn();
+        StarTreeIndexSpec starTreeIndexSpec = _indexingConfig.getStarTreeIndexSpec();
+        List<String> invertedIndexColumns = _indexingConfig.getInvertedIndexColumns();
+
+        // Check if the table config has any index configured
+        if ((sortedColumn != null && !sortedColumn.isEmpty()) || starTreeIndexSpec != null
+            || invertedIndexColumns != null) {
+          String indexGenerationOutputPath = _workingDir.getPath() + File.separator + INDEX_PREFIX + currentPartition;
+          try (
+              PinotSegmentRecordReader recordReader = new PinotSegmentRecordReader(outputSegment, null, sortedColumn)) {
+            buildSegment(indexGenerationOutputPath, _tableName, _segmentName, recordReader, _indexingConfig);
+          }
+          outputSegment = new File(indexGenerationOutputPath + File.separator + _segmentName);
+        }
+      }
+
+      resultFiles.add(outputSegment);
+    }
+    return resultFiles;
+  }
+
+  /**
+   * Helper function to trigger the segment creation
+   */
+  private void buildSegment(String outputPath, String tableName, String segmentName, RecordReader recordReader,
+      IndexingConfig indexingConfig) throws Exception {
+    SegmentGeneratorConfig segmentGeneratorConfig = new SegmentGeneratorConfig(recordReader.getSchema());
```

Comment:
This is not reading everything, so you would miss things like RAW index etc. This requires looking at segment metadata to derive all such info.